### PR TITLE
[Cal-3] Calendar scraper

### PR DESCRIPTION
**Pull request by CommanderStorm**
This PR adds a way to generate a SQLite database and keep it in sync with TUMonline.

(This PR is part of the #324 PR Train)

### CODE REVIEW

File: server/Cargo.toml [hunk 1]
```diff
@@ -18,9 +18,11 @@ strip = true
 
 [dependencies]
 log = "0.4.17"
-diesel = { version = "2.0.2", features = ["default","sqlite"] }
+diesel = { version = "2.0.2", features = ["default","chrono","sqlite"] }
 actix-web = "4.2.1"
-awc = "3.0.1"
+actix-rt = "2.7.0"
+rustls = "0.20.7"
+awc = { version= "3.0.1", features = ["rustls"] }
```
Comment (CommanderStorm) [inline server/Cargo.toml:25]: This is nessesary because we use https to communicate with TUMonline

File: server/load_api_data_to_db.py [hunk 2]
```diff
@@ -6,32 +6,51 @@
...
+            course_code             TEXT NULLABLE,
+            course_semester_hours   INTEGER NULLABLE,
+            group_id                TEXT NULLABLE,
+            xgroup                  TEXT NULLABLE,
+            status_id               TEXT NOT NULL,
+            status                  TEXT NOT NULL,
+            comment                 TEXT NOT NULL
+        );""",
+        )
+    # purposely, this index is only on this table and not on tmp_calendar
```
Comment (CommanderStorm) [inline server/load_api_data_to_db.py:52]:
```suggestion
    # purposely, this index is only on this table and not on calendar_scrape
```

File: server/src/calendar/continous_scraping.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    while !request_queue.is_empty() {
+        let mut new_request_queue = vec![];
+        for work in request_queue {
+            let events = XMLEvents::request(client, work.clone()).await;
+
+            //store the events in the database if successful, otherwise retry
+            match events {
+                Ok(events) => {
+                    success_cnt += events.len();
+                    events.store();
```
Comment (octycs) [inline server/src/calendar/continous_scraping.rs:260]: I think `store_in_db()` would be a bit better to understand

File: server/src/calendar/continous_scraping.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        assert_eq!(order.num_days(), 0);
+    }
+}
+
+async fn scrape(
+    client: &Client,
+    id: (String, i32),
+    from_year: i32,
+    year_duration: i32,
+) -> Result<usize, usize> {
```
Comment (octycs) [inline server/src/calendar/continous_scraping.rs:246]: Do I understand it correctly, that `Err` is only used to indicate that a retry was necessary, but not that the scraping failed? Maybe a custom `ScrapeResult` struct would be better then

File: server/src/calendar/continous_scraping.rs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+            .collect::<Vec<(String, i32)>>(),
+        Err(e) => {
+            error!("Error requesting all ids: {:?}", e);
+            vec![]
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct ScrapeOrder {
```
Comment (octycs) [inline server/src/calendar/continous_scraping.rs:141]: I'm not sure if I understand the name of this class. Do you mean "order" like an order in a restaurant or like the order of a sequence. Maybe `ScrapeTask` is better

Comment (CommanderStorm) [reply server/src/calendar/continous_scraping.rs:141]: True. Probably the explicit `ScrapeRoomTask` is even better.

File: server/src/calendar/continous_scraping.rs (hunk 3 above)
Comment (CommanderStorm) [reply server/src/calendar/continous_scraping.rs]: agree

File: server/src/calendar/continous_scraping.rs (hunk 4 above)
Comment (CommanderStorm) [reply server/src/calendar/continous_scraping.rs]: fully agree

**Review by octycs — APPROVED** [review]

**Pull request merged**